src/safeposix/syscalls/fs_calls.rs

```diff
@@ -910,6 +909,421 @@ impl Cage {
+        if let Some(inodenum) = metawalk(truepath.as_path(), Some(&metadata)) {
+            let thisinode = metadata.inodetable.get_mut(&inodenum).unwrap();
+            match thisinode {
+                Inode::File(ref mut general_inode) => {
+                    if mode & (S_IRWXA|(S_FILETYPEFLAGS as u32)) == mode {
+                        general_inode.mode = (general_inode.mode &!S_IRWXA) | mode
+                    } else {
+                        //there doesn't seem to be a good syscall error errno for this
+                        return syscall_error(Errno::EACCES, "chmod", "provided file mode is not valid");
+                    }
+                    //persist metadata
+                    return 0;
+                },
+                Inode::CharDev(ref mut dev_inode) => {
+                    if mode & (S_IRWXA|(S_FILETYPEFLAGS as u32)) == mode {
+                        dev_inode.mode = (dev_inode.mode &!S_IRWXA) | mode
+                    } else {
+                        //there doesn't seem to be a good syscall error errno for this
+                        return syscall_error(Errno::EACCES, "chmod", "provided file mode is not valid");
+                    }
+                    //persist metadata
+                    return 0;
+                },
+                Inode::Dir(ref mut dir_inode) => {
+                    if mode & (S_IRWXA|(S_FILETYPEFLAGS as u32)) == mode {
+                        dir_inode.mode = (dir_inode.mode &!S_IRWXA) | mode
+                    } else {
+                        //there doesn't seem to be a good syscall error errno for this
+                        return syscall_error(Errno::EACCES, "chmod", "provided file mode is not valid");
+                    }
+                    //persist metadata
+                    return 0;
+                },
+                Inode::Pipe(ref mut general_inode) => {
+                    if mode & (S_IRWXA|(S_FILETYPEFLAGS as u32)) == mode {
+                        general_inode.mode = (general_inode.mode &!S_IRWXA) | mode
+                    } else {
+                        //there doesn't seem to be a good syscall error errno for this
+                        return syscall_error(Errno::EACCES, "chmod", "provided file mode is not valid");
+                    }
+                    //persist metadata
+                    return 0;
+                },
+                Inode::Socket(ref mut general_inode) => {
+                    if mode & (S_IRWXA|(S_FILETYPEFLAGS as u32)) == mode {
+                        general_inode.mode = (general_inode.mode &!S_IRWXA) | mode
+                    } else {
+                        //there doesn't seem to be a good syscall error errno for this
+                        return syscall_error(Errno::EACCES, "chmod", "provided file mode is not valid");
+                    }
+                    //persist metadata
+                    return 0;
+                },
+            }
```

Comment:
If you persist metadata at the end of every arm, then you should just persist at the end instead

Reply:
Also return at the end, and error out on invalid mode  before this match, we want to duplicate as little code as possible